### PR TITLE
WIP: Copilot integration spike for creating and mapping routes

### PR DESCRIPTION
**Pull request by nmahoney-pivotal**
DO NOT MERGE THIS

Cloud Controller notifies Copilot on route creation and mapping. 

- We are submitting this early PR to get feedback
- There are no tests for what we added
- We have not considered deletion or unmapping but please let us know if you have suggestions

### CODE REVIEW

File: lib/cloud_controller.rb [hunk 1]
```diff
@@ -113,6 +113,7 @@ module VCAP::CloudController; end
 
 require 'cloud_controller/route_binding_message'
 require 'cloud_controller/process_route_handler'
+require 'cloud_controller/route_handler'
```
Comment (maryamklabib) [inline lib/cloud_controller.rb]: Note on naming:

slightly confusing to have `process_route_handler` and `route_handler`.

could we rename the latter to something that distinguishes it semantically? what does it handle routes for if not processes?

File: lib/cloud_controller/route_handler.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      )
+      logger.info("success upserting route in copilot")
+    rescue => e
+      logger.error("failed communicating with copilot server: #{e.message}")
+    end
+
+    private
+
+    def logger
+      @logger ||= Steno.logger('cc.process_route_handler')
```
Comment (maryamklabib) [inline lib/cloud_controller/route_handler.rb]: should this be `cc.route_handler` ?

File: app/controllers/runtime/routes_controller.rb [hunk 3]
```diff
@@ -160,9 +164,17 @@ def delete(guid)
...
         job = route_delete_action.delete_async(route: route, recursive: recursive_delete?)
+
+        # Are we messing up the response??
+        CopilotHandler.delete_route(guid) # if copilot.enabled?
+
         [HTTP::ACCEPTED, JobPresenter.new(job).to_json]
       else
         route_delete_action.delete_sync(route: route, recursive: recursive_delete?)
+
+        # Are we messing up the response??
```
Comment (maryamklabib) [inline app/controllers/runtime/routes_controller.rb]: What is the messing up that is feared?

File: app/actions/route_mapping_delete.rb [hunk 4]
```diff
@@ -22,6 +22,7 @@ def delete(route_mappings)
 
           route_mapping.destroy
           route_handler.update_route_information
+          CopilotHandler.new.unmap_route(route_mapping)
```
Comment (maryamklabib) [inline app/actions/route_mapping_delete.rb]: The instantiation and calling style of the `CopilotHandler` and the `ProcessRouteHandler` should be consistent.

Both should either be instantiated and called in the same line, or different lines.

@selzoc && maryam

File: app/actions/v2/route_mapping_create.rb [hunk 5]
```diff
@@ -44,6 +44,7 @@ def add
         RouteMappingModel.db.transaction do
           route_mapping.save
           route_handler.update_route_information
+          CopilotHandler.new.map_route(route_mapping) # if copilot.enabled?
```
Comment (maryamklabib) [inline app/actions/v2/route_mapping_create.rb]: How is the optionality of `copilot` going to be handled?

@selzoc && maryam

File: app/models/runtime/process_model.rb [hunk 6]
```diff
@@ -99,6 +99,7 @@ def buildpack_type
         [:spaces, :guid, :organization_id]
       ]
 
+    # Nick: seems important
```
Comment (maryamklabib) [inline app/models/runtime/process_model.rb]: Were these comments kept in by accident or did y'all have questions here?

@selzoc && maryam

File: lib/cloud_controller.rb [hunk 7]
```diff
@@ -113,6 +113,7 @@ module VCAP::CloudController; end
 
 require 'cloud_controller/route_binding_message'
 require 'cloud_controller/process_route_handler'
+require 'cloud_controller/copilot_handler'
```
Comment (maryamklabib) [inline lib/cloud_controller.rb]: naming is hard

some thoughts:
- `process_route_handler` handles process routes
- `copilot_handler` does not handle copilots
- how about `process_route_propogater` in addition to `process_route_handler`
- why not put copilot handler and client together?
- why not combine `copilot_handler` with `process_route_handler`?

File: lib/cloud_controller/dependency_locator.rb [hunk 8]
```diff
@@ -366,6 +370,16 @@ def build_bbs_stager_client
       VCAP::CloudController::Diego::BbsStagerClient.new(bbs_client)
     end
 
+    def build_copilot_client
+      Cloudfoundry::Copilot::Client.new(
+          host: 'copilot.service.cf.internal',
+          port: 9001,
```
Comment (maryamklabib) [inline lib/cloud_controller/dependency_locator.rb]: why this port? does it work cross releases?

**Comment by maryamklabib** [comment]
@nmahoney-pivotal @utako @Nino-K this may end up happening while cross team pairing. Do you plan to answer questions posed above? If you were seeking reactions, perhaps we can consider this closed?

**Comment by zrob** [comment]
this work now occurring as part of x-team pair.  wip pr is now dead

**Pull request closed without merging**